segments/vi_mode.p9k

```diff
@@ -4,18 +4,219 @@
+p9k::register_segment  "VI_MODE" "VISUAL" "${DEFAULT_COLOR}" "orange" ''  ''  ''  ''  ''
+
+################################################################
+# Register segment default values
+p9k::set_default P9K_VI_MODE_NORMAL_STRING 'NORMAL'
+p9k::set_default P9K_VI_MODE_INSERT_STRING 'INSERT'
+p9k::set_default P9K_VI_MODE_VISUAL_STRING 'VISUAL'
+
+###############################################################
 # Vi Mode: show editing mode (NORMAL|INSERT)
+# Parameters:
+#   * $1 Alignment: string - left|right
+#   * $2 Index: integer
+#   * $3 Joined: bool - If the segment should be joined
+################################################################
+# @description
+#   Display information about the editing mode - Normal | Insert | Visual.
+##
+# @args
+#   $1 string Alignment - left | right
+#   $2 integer Segment index
+#   $3 boolean Whether the segment should be joined
+##
+# @note
+#   Visual mode only works if the user has installed zsh-vimode-visual.
+##
+# @requires
+#   [zsh-vimode-visual](https://github.com/b4b4r07/zsh-vimode-visual)
+##
 prompt_vi_mode() {
-  case ${KEYMAP} in
-    vicmd)
-      "__p9k_$1_prompt_segment" "$0_NORMAL" "$2" "$DEFAULT_COLOR" "white" "$P9K_VI_COMMAND_MODE_STRING"
+  local vi_mode
+  local current_state
+  typeset -gAH vi_states
+  case "${KEYMAP}" in
+    vicmd)        current_state="NORMAL" ;;
+    vivis)        current_state="VISUAL" ;;
+    main|viins|*) current_state="INSERT" ;;
+  esac
+  vi_mode="P9K_VI_MODE_${current_state}_STRING"
+  # NOTE: The display test at the end of the next line ensures that the segment_state is displayed if an icon is defined
+  #       even if the string is set to an empty string. This allows you to define and use icons for this segment instead
+  #       of the default text string and also bypasses the ugly icon as content string. As a bonus, you can set both!
+  p9k::prepare_segment "$0" "${current_state}" $1 "$2" $3 "${(P)vi_mode}" "[[ '$__P9K_ICONS[VI_MODE_${current_state}]' != '' || '${(P)vi_mode}' != '' ]]"
+}
+
+###############################################################
+function rebuild_vi_mode {
+  case "${P9K_GENERATOR}" in
+    zsh-async)
+      if (( ${+terminfo[smkx]} )); then
+        printf '%s' ${terminfo[smkx]}
+      fi
+      local current_state
+      case "${KEYMAP}" in
+        vicmd)        current_state="NORMAL" ;;
+        vivis)        current_state="VISUAL" ;;
+        main|viins|*) current_state="INSERT" ;;
+      esac
+      local element joined=false
+      local vi_mode="P9K_VI_${current_state}_MODE_STRING"
+      for (( i = 0; i <= ${#P9K_LEFT_PROMPT_ELEMENTS}; i++ )); do
+        element="${(L)P9K_LEFT_PROMPT_ELEMENTS[$i]}"
+        if [[ "${element}" =~ "vi_mode" ]]; then
+          [[ "${element[-7,-1]}" == '_joined' ]] && joined=true
+          __p9k_async_callback "" "0" "VI_MODE_${current_state}·|·left·|·${i}·|·${joined}·|·${(P)vi_mode}·|·$__P9K_ICONS[VI_MODE_${current_state}]" "0" "0"
```

Comment:
I want to add the async generator in the next stage, but `zsh-async` needs to be pulled into `next` first. This code is there because I copied and pasted a lot from my testbed and probably missed this one.